### ISSUE TITLE
Re-factor all internally mutable structs to only be reference counted once

### ISSUE THREAD

**Issue by jennamagius** [oxy-secure/oxy#82]

**Comment by dcarosone** [comment]
I noticed a lot of `Rc<Refcell<..>>` fields all over, too, though I suddenly can't find as many as I remember, probably because I switched to the dev branch and you've already done some of this.

At some point tackling the shared state that leads to `RefCell` for interior mutability is probably also worthwhile, though it will be a challenge. All that runtime borrow checking has overhead, but more important is the potential for `panic`. A client can probably find ways to trigger them by sending syntactically-valid but semantically-unexpected messages, or something similar. Obviously it would be nice to minimise that potential, even if the impact is only DoS rather than something memory-unsafe.

This will be a big fight with the borrow checker and lifetimes, and avoiding that to get anything working (let alone something this functional in so little time) is more than enough justification. In the course of many refactorings, as roles and ownership of data become clearer, and more explicit error handling makes some cases statically verifiable by the compiler, some of this will naturally shake out. Eventually there will come time for a crusade (and extra tests) to get rid of more, or prove they're sound.

[All the above are clearly things you know, Jenna, but comments for posterity and future readers]

**Comment by jennamagius** [comment]
Well, the thing about a malicious client causing a server process to panic: that's not strictly a huge problem IMO? Server processes run in their own process, and panicing is a safe and acceptable thing to do.

In general I want to replace panics with graceful failures as a usability enhancement, but if you're an authed malicious party and you want to make your peer panic: go ahead and have fun.

**Comment by jennamagius** [comment]
The RefCells do literally have a runtime cost, but I extremely doubt they actually serve as a limiting reagent for practical performance. I'm completely in favor of cutting out unnecessary interior mutability where possible, because #128, but interior mutability also adds an EXTREME amount of expressive power, and it'd be a shame to make the codebase less nimble unless we're getting real value out of it.

One of the specific things that reference counted/interior mutable structs are good for is the "closure stack" pattern you see in copy::tick_remote_source. You send a StatRequest, then put a watch() closure on the response, then once you get the response you send a ReadDir and watch for THAT response, etc, etc.

That pattern is... pretty clumsy in a lot of ways. Like: you wind up writing `let proxy = proxy.clone()` every single layer of the closure stack. I think there's probably MANY ways to make that more elegant? And I'd love to hear ideas about that. But it's also a pattern that VERY expressively renders a six-message back-and-fourth exchange as a logically linear codeflow, and when you clear out interior mutability you largely take that option away from yourself.

**Comment by jennamagius** [comment]
Panicing is probably even more safe and acceptable if we build with `panic = "abort"`. Which we should probably do? We'd freeroll smaller binaries out of the transaction. I'd love to hear your thoughts.

**Comment by dcarosone** [comment]
panic is totally the right thing to do while in development and shaking things out. having a production service panic is not so good, but further points here:
* this is at least in part a deployment choice for the user/admin
* there's also a difference between panics in the core service and the per-user reexec'd server; as you note, the latter is probably still preferable, or at least acceptable for longer, so it helps prioritise where some of this work should be done first
* panics in client code might be different again; this is perhaps another (minor) point for separate binaries (#54), but again mostly about which parts of the code to focus on first.

And, yeah, if something panics it should be all over. All the more motivation to ensure they can't happen in places were they're not the right thing.

> One of the specific things that reference counted/interior mutable structs are good for is the "closure stack" pattern

Yep. Yep.  I come to Rust from Perl and AnyEvent, where callback closure chains are How Things Are Done (it's more-or-less Node.pl) and you put state scoped just above the chain to be captured by all the closures and added to along the way and GC'd at the end.

When futures finish shaking out, they might offer something here.

**Comment by jennamagius** [comment]
I would consider panicing the master pre-reexec server is a real and serious attack, you definitely should not be able to do that. There's not that much surface area for doing that though, that should be a pretty small thing to scrutinize.

Panicing a reexec process? I super don't care. You can just send Exit { } and get the same result.

Now, there is a distinction here between "Someone operating in good faith is liable to run into a panic" and "You can get a panic by being completely assinine". I definitely want to clear out the good faith cases, because that's just good user experience.

But finally: any situation where you get an AlreadyBorrowed panic is just a code error. Nothing should be holding on to a borrow for any non-trivial amount of time. But like the bug there isn't "We are using interior mutability", it's "you let the instruction pointer get away from you while you were still holding a borrow".
